src/client.rs

```diff
@@ -499,6 +502,100 @@ impl<T: Read + Write> Client<T> {
     }
 }
 
+#[cfg(feature = "rsasl")]
+impl<T: Read + Write> Client<T> {
+
+    /// Authenticate with the server using the given custom SASLConfig to handle the server's
+    /// challenge.
+    ///
+    pub fn sasl_auth(
+        mut self,
+        config: ::std::sync::Arc<SASLConfig>,
+        mechanism: &Mechname,
+    ) -> ::std::result::Result<Session<T>, (Error, Client<T>)> {
+        let client = SASLClient::new(config);
+        let session = ok_or_unauth_client_err!(
+            // TODO: Allow rsasl to select the mechanism
+            client.start_suggested(&[mechanism]).map_err(Error::AuthenticationSetup),
+            self
+        );
+        // TODO: check for `SASL-IR` capability and send initial data if it's supported.
+        ok_or_unauth_client_err!(
+            self.run_command(&format!("AUTHENTICATE {}", mechanism.as_str())),
+            self
+        );
+        self.do_sasl_handshake(session)
+    }
+
+    /// This func does the SASL handshake process once the authenticate command is made.
+    fn do_sasl_handshake(
```

Comment:
I wonder if we can pull out some of the code here into a helper function that can be used both by the existing `authenticate` and by the new SASL functionality. After all, they're doing _mostly_ the same thing.